### PR TITLE
SCP 939 PlayingFootstep Reimplementation

### PR DESCRIPTION
**Pull request by XKaguya**
Reimplementation from previous rep

### CODE REVIEW

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    internal static class PlayingFootstep
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label ret = generator.DefineLabel();
+
+            int offset = 2;
+            int index = newInstructions.FindLastIndex(i => i.opcode == OpCodes.Call && i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;
```
Comment (IRacle1) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:37]: as i see, offset should be `3`
```suggestion
            int offset = 3;
            int index = newInstructions.FindLastIndex(i => i.opcode == OpCodes.Call && i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;

```

Comment (IRacle1) [reply EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:37]: image

Comment (IRacle1) [reply EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:37]: also `i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))` is enough, it already check on `OpCodes.Call` opcode
```suggestion
            int offset = 2;
            int index = newInstructions.FindLastIndex(i => i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;

```

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                // Handlers.Scp939.OnPlayingFootstep(ev);
+                new(OpCodes.Call, Method(typeof(Scp939), nameof(Scp939.OnPlayingFootstep))),
+
+                // if (!IsAllowed)
+                //      return;
+                new(OpCodes.Callvirt, PropertyGetter(typeof(PlayingFootstepEventArgs), nameof(PlayingFootstepEventArgs.IsAllowed))),
+                new(OpCodes.Brfalse_S, ret),
+            });
+
+            newInstructions[newInstructions.Count - 1].WithLabels(ret);
```
Comment (IRacle1) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:62]: not a serious thing, but it is better to use `labels.Add` in situations where you dont need `WithLabels` return
```suggestion
            newInstructions[newInstructions.Count - 1].labels.Add(ret);

```

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    using PlayerRoles.FirstPersonControl.Thirdperson;
+    using PlayerRoles.PlayableScps.Scp939.Ripples;
+
+    using static HarmonyLib.AccessTools;
+
+    /// <summary>
+    /// Patches <see cref="FootstepRippleTrigger.OnFootstepPlayed(AnimatedCharacterModel, float)" />
+    /// to add the <see cref="Scp939.PlayingFootstep" /> event.
+    /// </summary>
+    [HarmonyPatch(typeof(FootstepRippleTrigger), nameof(FootstepRippleTrigger.OnFootstepPlayed))]
```
Comment (IRacle1) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:27]: also you should add event patch🤨

**Review by IRacle1 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs (hunk 3 above)
Comment (IRacle1) [reply EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:27]:
```cs
[EventPatch(typeof(Scp939), nameof(Scp939.PlayingFootstep))]
```

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs (hunk 1 above)
Comment (XKaguya) [reply EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs]: > as i see, offset should be `3`

fixed in the update, but if offset 2, which looks still works ? anyway

**Review by IRacle1 — APPROVED** [review]

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            int index = newInstructions.FindLastIndex(i => i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // Player player = Player.Get(hub);
+                new CodeInstruction(OpCodes.Ldarg_1).MoveLabelsFrom(newInstructions[index]),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(CharacterModel), nameof(CharacterModel.OwnerHub))),
+                new(OpCodes.Call, Method(typeof(API.Features.Player), nameof(API.Features.Player.Get), new[] { typeof(ReferenceHub) })),
+
+                // Player scp939 = Player.Get(hub);
```
Comment (Jesus-QC) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:48]:
```suggestion
                // Player scp939 = Player.Get(this.Player);

```

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label ret = generator.DefineLabel();
+
+            int offset = 3;
+            int index = newInstructions.FindLastIndex(i => i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // Player player = Player.Get(hub);
```
Comment (Jesus-QC) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:43]:
```suggestion
                // Player player = Player.Get(chatacterModel.OwnerHub);
```

**Review by Jesus-QC — CHANGES_REQUESTED** [review]

**Review by IRacle1 — DISMISSED** [review]

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            int index = newInstructions.FindLastIndex(i => i.Calls(Method(typeof(RippleTriggerBase), nameof(RippleTriggerBase.CheckVisibility)))) + offset;
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // Player target = Player.Get(characterModel.OwnerHub);
+                new CodeInstruction(OpCodes.Ldarg_1).MoveLabelsFrom(newInstructions[index]),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(CharacterModel), nameof(CharacterModel.OwnerHub))),
+                new(OpCodes.Call, Method(typeof(API.Features.Player), nameof(API.Features.Player.Get), new[] { typeof(ReferenceHub) })),
+
+                // Player player = Player.Get(this.Player);
```
Comment (Jesus-QC) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:48]:
```suggestion
                // Player player = Player.Get(this.Owner);
```

File: EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Callvirt, PropertyGetter(typeof(RippleTriggerBase), nameof(RippleTriggerBase.Player))),
```
Comment (Jesus-QC) [inline EXILED/Exiled.Events/Patches/Events/Scp939/PlayingFootstep.cs:50]:
```suggestion
                new(OpCodes.Callvirt, PropertyGetter(typeof(RippleTriggerBase), nameof(RippleTriggerBase.Owner))),
```

**Review by Vladislav-CS — APPROVED** [review]

**Review by Jesus-QC — APPROVED** [review]

**Review by IRacle1 — APPROVED** [review]
ГОЙДА

**Pull request merged**